distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,856 @@
+        return unparse_host_port(ensure_ip(host), port)
+
+    def get_local_address_for(self, loc):
+        host, port = parse_host_port(loc)
+        host = ensure_ip(host)
+        if ":" in host:
+            local_host = get_ipv6(host)
+        else:
+            local_host = get_ip(host)
+        return unparse_host_port(local_host, None)
+
+
+class TLSBackend(TCPBackend):
+    _connector_class = TLSConnector
+    _listener_class = TLSListener
+
+
+class _ZeroCopyWriter:
+    """The builtin socket transport in asyncio makes a bunch of copies, which
+    can make sending large amounts of data much slower. This hacks around that.
+
+    Note that this workaround isn't used with the windows ProactorEventLoop or
+    uvloop."""
+
+    SENDMSG_MAX_SIZE = 1024 * 1024  # 1 MiB
+    SENDMSG_MAX_COUNT = 16
+
+    def __init__(self, transport):
+        self.transport = transport
+        self._buffers = collections.deque()
+        self._offset = 0
+
+    def _buffer_append(self, data):
+        self._buffers.append(memoryview(data).cast("B"))
+
+    def _buffer_peek(self):
+        offset = self._offset
+        buffers = []
+        size = 0
+        count = 0
+        for b in self._buffers:
+            if offset:
+                b = b[offset:]
+                offset = 0
+            buffers.append(b)
+            size += len(b)
+            count += 1
+            if size > self.SENDMSG_MAX_SIZE or count == self.SENDMSG_MAX_COUNT:
+                break
+        return buffers
+
+    def _buffer_advance(self, size):
+        offset = self._offset
+        buffers = self._buffers
+        while size:
+            b = buffers[0]
+            b_len = len(b) - offset
+            if b_len <= size:
+                buffers.popleft()
+                size -= b_len
+                offset = 0
+            else:
+                offset += size
+                break
+        self._offset = offset
+
+    def write(self, data):
+        transport = self.transport
+
+        if transport._eof:
+            raise RuntimeError("Cannot call write() after write_eof()")
+        if transport._empty_waiter is not None:
+            raise RuntimeError("unable to write; sendfile is in progress")
+        if not data:
+            return
+        if transport._conn_lost:
+            return
+
+        if not self._buffers:
+            try:
+                n = transport._sock.send(data)
+            except (BlockingIOError, InterruptedError):
+                pass
+            except (SystemExit, KeyboardInterrupt):
+                raise
+            except BaseException as exc:
+                transport._fatal_error(exc, "Fatal write error on socket transport")
+                return
+            else:
+                data = data[n:]
+                if not data:
+                    return
+            # Not all was written; register write handler.
+            transport._loop._add_writer(transport._sock_fd, self._on_write_ready)
```

Comment:
We need to use the private one since we're mucking with asyncio internals. The public one includes a check that the file descriptor being watched doesn't correspond to an active transport (presumably to stop a user from shooting themselves in the foot). I want to shoot myself in the foot, so we use the private method :).